### PR TITLE
chore: Enables mongodbatlas_push_based_log_export tests in CI and adds unit tests

### PR DESCRIPTION
**Pull request by maastha**
## Description

Enables mongodbatlas_push_based_log_export tests in CI and adds unit tests

Link to any related issue(s): https://jira.mongodb.org/browse/CLOUDP-240975

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR. A migration guide must be created or updated if the new feature will go in a major version.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR. A migration guide must be created or updated.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contributing guides](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/contributing/README.md)
- [ ] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [ ] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [ ] I have added any necessary documentation (if appropriate)
- [ ] I have run make fmt and formatted my code
- [ ] If changes include deprecations or removals, I defined an isolated PR with a relevant title as it will be used in the auto-generated changelog.
- [ ] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

### CODE REVIEW

**Comment by lantoli** [comment]
tests are passing but showing "400 (request "CANNOT_ASSUME_ROLE") Atlas cannot assume the specified role" warning, don't know if it's important

File: .github/workflows/acceptance-tests-runner.yml [hunk 1]
```diff
@@ -154,6 +154,7 @@ jobs:
       search_index: ${{ steps.filter.outputs.search_index == 'true' || env.mustTrigger == 'true' }}
       serverless: ${{ steps.filter.outputs.serverless == 'true' || env.mustTrigger == 'true' }}
       stream: ${{ steps.filter.outputs.stream == 'true' || env.mustTrigger == 'true' }}
+      push_based_log_export: ${{ steps.filter.outputs.push_based_log_export == 'true' || env.mustTrigger == 'true' }}
```
Comment (lantoli) [inline .github/workflows/acceptance-tests-runner.yml]: can you order this as well?

File: .github/workflows/acceptance-tests-runner.yml [hunk 2]
```diff
@@ -234,6 +235,8 @@ jobs:
           stream:
             - 'internal/service/streamconnection/*.go'
             - 'internal/service/streaminstance/*.go'
+          push_based_log_export:
```
Comment (lantoli) [inline .github/workflows/acceptance-tests-runner.yml]: also order this

File: internal/service/pushbasedlogexport/model_test.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+	timeout         *timeouts.Value
+	expectedTFModel *pushbasedlogexport.TFPushBasedLogExportRSModel
+	name            string
+	projectID       string
+}
+
+func TestNewTFPushBasedLogExport(t *testing.T) {
+	currentTime := time.Now()
+
+	testCases := []sdkToTFModelTestCase{
```
Comment (lantoli) [inline internal/service/pushbasedlogexport/model_test.go:37]: knit: i think it's a bit better using a string map for the test name, e.g.: https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/internal/service/project/model_project_test.go#L550-L557

File: internal/service/pushbasedlogexport/model_test.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+	expectedUpdateReq *admin.PushBasedLogExportProject
+	name              string
+}
+
+func TestNewPushBasedLogExportReq(t *testing.T) {
+	testCases := []pushBasedLogExportReqTestCase{
+		{
+			name: "Valid TF state",
+			input: &pushbasedlogexport.TFPushBasedLogExportRSModel{
+				BucketName: types.StringValue(testBucketName),
```
Comment (lantoli) [inline internal/service/pushbasedlogexport/model_test.go:100]: to simplify test we might want to have in testBucketName the value of types.StringValue(testBucketName) so we can have just: BucketName: testBucketName,

same for all

File: internal/service/pushbasedlogexport/state_transition_test.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+var (
+	activeState           = "ACTIVE"
+	unconfiguredState     = "UNCONFIGURED"
+	initiatingState       = "INITIATING"
+	bucketVerifiedState   = "BUCKET_VERIFIED"
+	assumeRoleFailedState = "ASSUME_ROLE_FAILED"
+	unknown               = ""
+	sc500                 = conversion.IntPtr(500)
+	currentTime           = time.Now()
```
Comment (lantoli) [inline internal/service/pushbasedlogexport/state_transition_test.go:29]: i think it's ok here but sometimes using the current (real) time can lead to flaky tests

File: internal/testutil/acc/name.go [hunk 6]
```diff
@@ -45,3 +46,7 @@ func RandomEmail() string {
 func RandomLDAPName() string {
 	return fmt.Sprintf("CN=%s-%s@example.com,OU=users,DC=example,DC=com", prefixName, acctest.RandString(10))
 }
+
+func RandomS3BucketName() string {
```
Comment (lantoli) [inline internal/testutil/acc/name.go:50]: nice!

**Review by lantoli — APPROVED** [review]

**Review by oarbusi — APPROVED** [review]

File: internal/service/pushbasedlogexport/model_test.go [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+				State:      types.StringValue(activeState),
+				CreateDate: types.StringPointerValue(conversion.TimePtrToStringPtr(&currentTime)),
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			resultModel, _ := pushbasedlogexport.NewTFPushBasedLogExport(context.Background(), tc.projectID, tc.apiResp, tc.timeout)
+			if !reflect.DeepEqual(resultModel, tc.expectedTFModel) {
```
Comment (EspenAlbert) [inline internal/service/pushbasedlogexport/model_test.go]: Why not `assert.Equal(t, resultModel, tc.expectedTFModel)` ?

File: internal/testutil/acc/name.go [hunk 8]
```diff
@@ -7,11 +7,12 @@ import (
...
-	prefixProject = prefixName + "-p"
-	prefixCluster = prefixName + "-c"
-	prefixIAMRole = "mongodb-atlas-" + prefixName
-	prefixIAMUser = "arn:aws:iam::358363220050:user/mongodb-aws-iam-auth-test-user"
+	prefixName     = "test-acc-tf"
+	prefixProject  = prefixName + "-p"
+	prefixCluster  = prefixName + "-c"
+	prefixIAMRole  = "mongodb-atlas-" + prefixName
+	prefixIAMUser  = "arn:aws:iam::358363220050:user/mongodb-aws-iam-auth-test-user"
+	prefixS3Bucket = "mongodb-atlas"
```
Comment (EspenAlbert) [inline internal/testutil/acc/name.go]: Worth mentioning `mongodb-atlas-tf`?

File: internal/service/pushbasedlogexport/model_test.go [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+			expectedCreateReq: &admin.CreatePushBasedLogExportProjectRequest{
+				BucketName: testBucketName,
+				IamRoleId:  testIAMRoleID,
+				PrefixPath: prefixPathEmpty,
+			},
+			expectedUpdateReq: &admin.PushBasedLogExportProject{
+				BucketName: admin.PtrString(testBucketName),
+				IamRoleId:  admin.PtrString(testIAMRoleID),
+				PrefixPath: admin.PtrString(prefixPathEmpty),
+			},
```
Comment (AgustinBettati) [inline internal/service/pushbasedlogexport/model_test.go:131]: nice having both request models asserted under the same test

**Review by AgustinBettati — APPROVED** [review]
LGTM

File: internal/service/pushbasedlogexport/model_test.go (hunk 7 above)
Comment (maastha) [reply internal/service/pushbasedlogexport/model_test.go]: updated

File: internal/testutil/acc/name.go (hunk 8 above)
Comment (maastha) [reply internal/testutil/acc/name.go]: updated

File: .github/workflows/acceptance-tests-runner.yml (hunk 2 above)
Comment (maastha) [reply .github/workflows/acceptance-tests-runner.yml]: updated

File: .github/workflows/acceptance-tests-runner.yml (hunk 1 above)
Comment (maastha) [reply .github/workflows/acceptance-tests-runner.yml]: updated

File: internal/service/pushbasedlogexport/state_transition_test.go (hunk 5 above)
Comment (maastha) [reply internal/service/pushbasedlogexport/state_transition_test.go:29]: let me know if you have any suggestions? I can update that if needed

File: internal/service/pushbasedlogexport/model_test.go (hunk 4 above)
Comment (maastha) [reply internal/service/pushbasedlogexport/model_test.go:100]: I don't see a lot of value in doing that, we're not only using types.StringValue(testBucketName) but also admin.PtrString() and the strings as is

**Comment by maastha** [comment]
>tests are passing but showing "400 (request "CANNOT_ASSUME_ROLE") Atlas cannot assume the specified role" warning, don't know if it's important

yes this is expected with the cloud provider access resource API. The resource then does retries to get a successful response.

**Pull request merged**